### PR TITLE
Add stable-7 commits to api-8

### PR DESCRIPTION
**Pull request by ImMorpheus**
Haven't checked everything but this is a start. I'm opening a wip PR because it's not finished but what's pushed can be reviewed.

### Now for the fun part

There are:
missing commits from api-8 history which can be cherry-picked (this PR)
missing commits from api-8 history which cannot be cherry-picked (the old api has been nuked, command api is a good example of this). This means bugs fixed by those commits needs to be retested on the new API (out of scope for this PR)
missing commits from api-8 history which cannot be cherry-picked because their classes are missing (read this as "the world API is gone, reduced to atoms") 

### Be more specific!

https://github.com/SpongePowered/SpongeAPI/commit/c154fd4b06cddc1dd0f3557f46349eddf681fee1
https://github.com/SpongePowered/SpongeAPI/commit/bb14b831b537a7b37539eaf6fd3fa72c5090f9c1
https://github.com/SpongePowered/SpongeAPI/commit/bd284c2285db20792bddc554615b9fefcf0e8986
https://github.com/SpongePowered/SpongeAPI/commit/4e17b0a7b2ae122d98494e13cc058e5f09f0ec03
https://github.com/SpongePowered/SpongeAPI/commit/96bb9d1b806a4ac8758d53f2784f4acf5fe66218
https://github.com/SpongePowered/SpongeAPI/commit/f12f8e7743844b8761b890ce3e440911e97c6f38

### But wait, there's more

https://github.com/SpongePowered/SpongeAPI/commit/85a951a7c250110a19ece0276cf218bb1c9a1d05

That commit was not completely cherry-picked.
`ChangeLevelEvent` is deprecated on api7, but perfectly fine on api8 (fixed in this PR)
`ChangeEntityExperienceEvent` has all of its int-based methods deprecated in favour of Data-based one on api-7, but no deprecation or data methods on api8. What's more, that event is now screwed because there's no `ExperienceHolderData` in api8 so it cannot be cherry-picked.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/placeholder/PlaceholderParsers.java [hunk 1]
```diff
@@ -22,47 +22,32 @@
...
 
     /**
-     * Gets the original level of the human.
-     *
-     * @return The original level of the human
+     * A parser that returns the associated source's current world, if
+     * applicable, else the default world.
      */
-    int getOriginalLevel();
+    static PlaceholderParser CURRENT_WORLD = DummyObjectProvider.createFor(PlaceholderParser.class, "CURRENT_WORLD");
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/placeholder/PlaceholderParsers.java]: Should turn into Suppliers, `DummyObjectProvider` is gonna be nuked.

**Review by gabizou — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/api/placeholder/PlaceholderParsers.java (hunk 1 above)
Comment (ImMorpheus) [reply src/main/java/org/spongepowered/api/placeholder/PlaceholderParsers.java]: Done

**Comment by dualspiral** [comment]
Is it possible to get this done soon please? I've just hit a point porting Nucleus where I need something currently not in API 8 (placeholders) so I personally would like it (but it would be good to get this off the plate, regardless!)

**Comment by ImMorpheus** [comment]
Done

**Comment by dualspiral** [comment]
@gabizou Does this look good to you?

File: src/main/java/org/spongepowered/api/CatalogTypes.java [hunk 2]
```diff
@@ -220,6 +221,9 @@
 
     public static final Class<PortalType> PORTAL_TYPE = PortalType.class;
 
+    public static final Class<PlaceholderParser> PLACEHOLDER_PARSER = PlaceholderParser.class;
+
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/CatalogTypes.java:225]: Extra lines

File: src/main/java/org/spongepowered/api/placeholder/PlaceholderParsers.java [hunk 3]
```diff
@@ -22,47 +22,34 @@
...
+
+import java.util.function.Supplier;
 
 /**
- * Called when a human's level is changed.
+ * Contains Sponge provided {@link PlaceholderParser}s.
  */
-@GenerateFactoryMethod
-public interface ChangeLevelEvent extends Event, Cancellable {
+public class PlaceholderParsers {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/placeholder/PlaceholderParsers.java]: finalize

File: src/main/java/org/spongepowered/api/placeholder/PlaceholderText.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+ *
+ * <p>The {@link PlaceholderContext} is fixed when this object is created, but
+ * {@link PlaceholderParser#parse(PlaceholderContext)} is not called until
+ * {@link #asComponent()} is called. Thus, any {@link TextComponent} object that is created
+ * will reflect the time that the {@link TextComponent} object was requested, and not when
+ * this object itself was created. It therefore follows that implementations must
+ * not cache the result of {@link #asComponent()} unless it is known that the supplied
+ * parser is <strong>not</strong> sensitive to the time of invocation.</p>
+ */
+public interface PlaceholderText extends ComponentLike {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/placeholder/PlaceholderText.java]: PlaceholderComponent

**Pull request merged**